web/server/codechecker_server/api/authentication.py

```diff
@@ -167,6 +206,85 @@ def performLogin(self, auth_method, auth_string):
                     codechecker_api_shared.ttypes.ErrorCode.AUTH_DENIED,
                     msg)
 
+        elif auth_method == "oauth":
+            provider, url = auth_string.split("@")
+
+            oauth_config = self.__manager.get_oauth_config(provider)
+            if not oauth_config.get('enabled'):
+                LOG.error("OAuth authentication is " +
+                          "not enabled for provider: %s", provider)
+                raise codechecker_api_shared.ttypes.RequestFailed(
+                    codechecker_api_shared.ttypes.ErrorCode.AUTH_DENIED,
+                    "OAuth authentication is not enabled.")
+
+            client_id = oauth_config["oauth_client_id"]
+            client_secret = oauth_config["oauth_client_secret"]
+            scope = oauth_config["oauth_scope"]
+            token_url = oauth_config["oauth_token_uri"]
+            user_info_url = oauth_config["oauth_user_info_uri"]
+            redirect_uri = oauth_config["oauth_redirect_uri"]
+            allowed_users = oauth_config.get("allowed_users", [])
+
+            LOG.info("OAuth configuration loaded for provider: %s", provider)
+
+            session = OAuth2Session(
+                client_id,
+                client_secret,
+                scope=scope,
+                redirect_uri=redirect_uri)
+
+            # FIXME: This is a workaround for the Microsoft OAuth2 provider
+            # which doesn't correctly fetch the code from url.
+
+            url_new = urlparse(url)
+            parsed_query = parse_qs(url_new.query)
+
+            code = parsed_query.get("code")[0]
+            state = parsed_query.get("state")[0]
+
+            url = url_new.scheme + "://" + url_new.netloc + url_new.path + \
+                "?code=" + code + "&state=" + state
+
+            LOG.info("Constructed URL")
+
+            token = session.fetch_token(
+                url=token_url,
+                authorization_response=url)
+
+            LOG.info("Token fetched successfully for provider: %s", provider)
+
+            user_info = session.get(user_info_url).json()
+            username = user_info[
+                oauth_config["oauth_user_info_mapping"]["username"]]
+            LOG.info("User info fetched, username: %s", username)
+
+            if provider == "github" and \
+                    "localhost" not in user_info_url:
+                link = "https://api.github.com/user/emails"
+                for email in session.get(link).json():
+                    if email['primary']:
+                        username = email['email']
+                        LOG.info("Primary email found: %s", username)
+                        break
+
+            if allowed_users == ["*"] or username in allowed_users:
+                LOG.info("User %s is authorized.", username)
+                session = self.__manager.create_session(
+                    f"{provider}@{username}:{token['access_token']}")
+                return session.token
```

Comment:
What happens when an OAuth user changes their password on the OAuth service? Or their account gets flagged for fraud or similar? According to the OAuth service their account will be disabled and the `access_token` revoked, but on the CodeChecker service they will stay logged in.
The session manager must store the `access_token` associated with the session, and periodically check the validity of the stored `access_token` on user interaction. It must also handle refreshing it using the `refresh_token`, if the `access_token` expired. (These could both be handled by authlib.)